youtube_dl/extractor/mixcloud.py

```diff
@@ -84,54 +69,108 @@ def _real_extract(self, url):
+                key = self._search_regex(r'{0}({1}[^{0}]*){0}'.format(quote, re.escape(partial_key)), js,
+                                         "encryption key", default=None)
+                if key is not None:
+                    break
+            else:
+                continue
+            break
+        else:
+            raise ExtractorError('Failed to extract encryption key')
+
+        if encrypted_play_info is not None:
+            play_info = self._parse_json(self._decrypt_xor_cipher(key, encrypted_play_info), 'play info')
+            if message and 'stream_url' not in play_info:
+                raise ExtractorError('%s said: %s' % (self.IE_NAME, message), expected=True)
+            song_url = play_info['stream_url']
+            formats = [{
+                'format_id': 'normal',
+                'url': song_url
+            }]
+
+            title = self._html_search_regex(r'm-title="([^"]+)"', webpage, 'title')
+            thumbnail = self._proto_relative_url(self._html_search_regex(
+                r'm-thumbnail-url="([^"]+)"', webpage, 'thumbnail', fatal=False))
+            uploader = self._html_search_regex(
+                r'm-owner-name="([^"]+)"', webpage, 'uploader', fatal=False)
+            uploader_id = self._search_regex(
+                r'\s+"profile": "([^"]+)",', webpage, 'uploader id', fatal=False)
+            description = self._og_search_description(webpage)
+            view_count = str_to_int(self._search_regex(
+                [r'<meta itemprop="interactionCount" content="UserPlays:([0-9]+)"',
+                 r'/listeners/?">([0-9,.]+)</a>',
+                 r'(?:m|data)-tooltip=["\']([\d,.]+) plays'],
+                webpage, 'play count', default=None))
+
+        else:
+            title = info_json['name']
+            thumbnail = try_get(info_json,
+                                lambda x: 'https://thumbnailer.mixcloud.com/unsafe/600x600/' + x['picture']['urlRoot'])
+            uploader = try_get(info_json, lambda x: x['owner']['displayName'])
+            uploader_id = try_get(info_json, lambda x: x['owner']['username'])
+            description = try_get(info_json, lambda x: x['description'])
+            view_count = try_get(info_json, lambda x: x['plays'])
+            formats = [{
+                'format_id': 'normal',
+                'url': self._decrypt_xor_cipher(key, base64.b64decode(info_json['streamInfo']['url']))
+            }]
+
+            hls_encrypted = try_get(info_json, lambda x: x['streamInfo']['hlsUrl'])
+            if hls_encrypted is not None:
+                hls_url = self._decrypt_xor_cipher(key, base64.b64decode(hls_encrypted))
+                formats.extend(self._extract_m3u8_formats(hls_url, title))
+
+            dash_encrypted = try_get(info_json, lambda x: x['streamInfo']['dashUrl'])
+            if dash_encrypted is not None:
+                dash_url = self._decrypt_xor_cipher(key, base64.b64decode(dash_encrypted))
```

Comment:
157, 160-162, 165-167 code duplication.

Reply:
What would you suggest then? The duplication is minimal.

Reply:
Extracting duplicate code into a function obviously.